eras/conway/impl/src/Cardano/Ledger/Conway/Rules/Deleg.hs

```diff
@@ -0,0 +1,219 @@
+      !(StrictMaybe Coin)
+  | StakeKeyAlreadyRegisteredDELEG
+      !(Credential 'Staking (EraCrypto era))
+  | StakeKeyNotRegisteredDELEG
+      !(Credential 'Staking (EraCrypto era))
+  | StakeKeyHasNonZeroAccountBalanceDELEG
+      !(Maybe Coin)
+  | DRepAlreadyRegisteredForStakeKeyDELEG
+      !(Credential 'Staking (EraCrypto era))
+  | WrongCertificateTypeDELEG
+  deriving (Show, Eq, Generic)
+
+instance NoThunks (ConwayDelegPredFailure era)
+
+instance NFData (ConwayDelegPredFailure era)
+
+instance Era era => EncCBOR (ConwayDelegPredFailure era) where
+  encCBOR =
+    encode . \case
+      IncorrectDepositDELEG mCoin ->
+        Sum (IncorrectDepositDELEG @era) 1
+          !> To mCoin
+      StakeKeyAlreadyRegisteredDELEG stakeCred ->
+        Sum (StakeKeyAlreadyRegisteredDELEG @era) 2
+          !> To stakeCred
+      StakeKeyNotRegisteredDELEG stakeCred ->
+        Sum (StakeKeyNotRegisteredDELEG @era) 3
+          !> To stakeCred
+      StakeKeyHasNonZeroAccountBalanceDELEG mCoin ->
+        Sum (StakeKeyHasNonZeroAccountBalanceDELEG @era) 4
+          !> To mCoin
+      DRepAlreadyRegisteredForStakeKeyDELEG stakeCred ->
+        Sum (DRepAlreadyRegisteredForStakeKeyDELEG @era) 5
+          !> To stakeCred
+      WrongCertificateTypeDELEG ->
+        Sum (WrongCertificateTypeDELEG @era) 6
+
+instance Era era => DecCBOR (ConwayDelegPredFailure era) where
+  decCBOR = decode $ Summands "ConwayDelegPredFailure" $ \case
+    1 -> SumD IncorrectDepositDELEG <! From
+    2 -> SumD StakeKeyAlreadyRegisteredDELEG <! From
+    3 -> SumD StakeKeyNotRegisteredDELEG <! From
+    4 -> SumD StakeKeyHasNonZeroAccountBalanceDELEG <! From
+    5 -> SumD DRepAlreadyRegisteredForStakeKeyDELEG <! From
+    6 -> SumD WrongCertificateTypeDELEG
+    n -> Invalid n
+
+newtype ConwayDelegEvent era = DelegEvent (Event (EraRule "DELEG" era))
+
+instance
+  ( EraPParams era
+  , State (EraRule "DELEG" era) ~ DState era
+  , Signal (EraRule "DELEG" era) ~ ConwayDelegCert (EraCrypto era)
+  , Environment (EraRule "DELEG" era) ~ DelegEnv era
+  , EraRule "DELEG" era ~ ConwayDELEG era
+  ) =>
+  STS (ConwayDELEG era)
+  where
+  type State (ConwayDELEG era) = DState era
+  type Signal (ConwayDELEG era) = ConwayDelegCert (EraCrypto era)
+  type Environment (ConwayDELEG era) = DelegEnv era
+  type BaseM (ConwayDELEG era) = ShelleyBase
+  type PredicateFailure (ConwayDELEG era) = ConwayDelegPredFailure era
+  type Event (ConwayDELEG era) = ConwayDelegEvent era
+
+  transitionRules = [conwayDelegTransition @era]
+
+conwayDelegTransition :: forall era. EraPParams era => TransitionRule (ConwayDELEG era)
+conwayDelegTransition = do
+  TRC
+    ( DelegEnv _slot _ptr _acnt pp
+      , dState@DState {dsUnified}
+      , c
+      ) <-
+    judgmentContext
+  let pd = pp ^. ppKeyDepositL
+  case c of
+    ConwayRegCert stakeCred sMayDeposit -> do
+      checkDepositAgainstPParams pd sMayDeposit
+      checkStakeKeyNotAlreadyRegistered stakeCred dsUnified
```

Comment:
It might also be delegating to a DRep. I'm not completely sure what we want to do about the other things it might be delegated to (e.g. Ptrs).